### PR TITLE
Factor refactoring

### PR DESCRIPTION
**Pull request by francescoTheSantis**
Mid-level refactoring:

1. Variable - the information regarding the variable parents is not specified in the variable anymore.
2. ParametricFactor - a base factor composed by just a scope (the set of variables over which it operates) and neural network parameterising the factor.
3. ParametricCPD - a directed factor (subclass of ParametricFactor) with explicit parents, representing a conditional probability distribution. 
4. ProbabilisticModel - the unified container. Pass ParametricCPDs to get a directed model (Bayesian Network). Pass ParametricFactors for an undirected probabilistic model.
5. Inference modules (ForwardInference, AncestralSamplingInference, IndependentInference, DeterministicInference) - updated given the changes 1-4.
6. Constructors (GraphModel, BipartiteModel) - updated given the changes 1-4.
7. The relative examples have also been updated with for new API.

### CODE REVIEW

File: CONTRIBUTOR_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,944 @@
+# PyC (PyTorch Concepts) — Comprehensive Contributor Guide
```
Comment (pietrobarbiero) [inline CONTRIBUTOR_GUIDE.md]: This commit doesn't seem related to the purpose of the PR and the content might also change a lot

File: torch_concepts/nn/modules/mid/models/probabilistic_model.py [hunk 2]
```diff
@@ -1,120 +1,119 @@
...
+    ------
+    TypeError
+        If neither ``factors`` nor ``parametric_cpds`` is provided, or if
+        both are provided simultaneously.
     """
-    def __init__(self, variables: List[Variable], parametric_cpds: List[ParametricCPD]):
+
+    def __init__(self, variables: List[Variable],
+                 factors: List[ParametricFactor] = None,
+                 parametric_cpds: List[ParametricCPD] = None):
```
Comment (pietrobarbiero) [inline torch_concepts/nn/modules/mid/models/probabilistic_model.py]: can't we just have 2 lists in the constructor? variable and factors (parametricCPDs are factors...)
the distinction can be checked internally

**Review by pietrobarbiero — CHANGES_REQUESTED** [review]
Please double check that we could skip redundant argument in the ProbabilisticModel constructor

**Comment by francescoTheSantis** [comment]
The requested change has been handled. Now the ProbabilisticModel is instantiated by defining a list of variables and a list of factors (ParametricFactors or ParametricCPDs). Then the ProbabilisticModel internally checks if factors contains either ParametricFactors or ParametricCPDs and proceeds accordingly (e.g., defining the topological order if ParametricCPDs).

File: torch_concepts/nn/modules/mid/models/probabilistic_model.py [hunk 3]
```diff
@@ -1,120 +1,115 @@
...
+    Raises
+    ------
+    TypeError
+        If the factors list mixes :class:`ParametricCPD` and plain
+        :class:`ParametricFactor` instances.
     """
-    def __init__(self, variables: List[Variable], parametric_cpds: List[ParametricCPD]):
+
+    def __init__(self, variables: List[Variable],
+                 factors: List[ParametricFactor]):
```
Comment (gdefe) [inline torch_concepts/nn/modules/mid/models/probabilistic_model.py]: we should have better typing here describing all possible input types.

maybe Union[List[ParametricFactor], List[ParametricCPD]]

File: examples/utilization/1_pgm/0_concept_bottleneck_model.py [hunk 4]
```diff
@@ -25,25 +25,25 @@ def main():
...
+    concept_model = ProbabilisticModel(variables=[input_var, *concepts, tasks], factors=[backbone, *c_encoder, y_predictor])
 
     # Inference Initialization
     inference_engine = DeterministicInference(concept_model)
     initial_input = {'input': x_train}
     query_concepts = ["c1", "c2", "xor"]
 
     optimizer = torch.optim.AdamW(concept_model.parameters(), lr=0.01)
-    loss_fn = torch.nn.BCEWithLogitsLoss()
+    loss_fn = torch.nn.BCELoss()
```
Comment (gdefe) [inline examples/utilization/1_pgm/0_concept_bottleneck_model.py]: why changing to not Logits? the layers output logits if not activated by the inference

File: examples/utilization/1_pgm/0_concept_bottleneck_model.py [hunk 5]
```diff
@@ -62,8 +62,8 @@ def main():
         optimizer.step()
 
         if epoch % 100 == 0:
-            task_accuracy = accuracy_score(y_train, y_pred > 0.)
-            concept_accuracy = accuracy_score(c_train, c_pred > 0.)
+            task_accuracy = accuracy_score(y_train, y_pred.detach() > 0.5)
```
Comment (gdefe) [inline examples/utilization/1_pgm/0_concept_bottleneck_model.py]: same here

File: torch_concepts/nn/modules/mid/models/cpd.py [hunk 6]
```diff
@@ -308,6 +278,3 @@ def build_potential(self) -> torch.Tensor:
             raise NotImplementedError("Potential table construction not supported for this distribution.")
 
         return potential_table
-
```
Comment (gdefe) [inline torch_concepts/nn/modules/mid/models/cpd.py:311]: why removing the __repr__?

Comment (gdefe) [reply torch_concepts/nn/modules/mid/models/cpd.py:311]: got it, you have a common one in Factor

**Comment by francescoTheSantis** [comment]
I updated the typing in the ProbabilistiModel (factors = Union[List[ParametricFactor], List[ParametricCPD]]).
In the example script 0_concept_bottleneck_model.py now the query is executed using return_logits=True, which returns logit values.
